### ISSUE TITLE
upload blob does not seem to use timeoutIntervalInMs

### ISSUE THREAD

**Issue by dadoonet** [Azure/azure-storage-java#71]
I read the code and it sounds like to me that when you upload a blob on azure storage, `timeoutIntervalInMs` is never used.

Same is happening when using `getOutputStream()` or `getInputStream()`.

```java
public void testDumb() throws URISyntaxException, StorageException {
    CloudStorageAccount storageAccount = CloudStorageAccount.getDevelopmentStorageAccount();
    CloudBlobClient client = storageAccount.createCloudBlobClient();
    client.getDefaultRequestOptions().setTimeoutIntervalInMs(1);
    client.getContainerReference("dumb").getBlockBlobReference("blob").upload(something, 10000);
}
```

Could you confirm my understanding? And if I'm right, is there any way to fix that?

See https://github.com/elastic/elasticsearch/issues/12567#issuecomment-171933503 for details.

**Comment by emgerner-msft** [comment]
Where are you reading this in the code? You can see timeout is added to every single service request made in the library [here](https://github.com/Azure/azure-storage-java/blob/master/microsoft-azure-storage/src/com/microsoft/azure/storage/core/BaseRequest.java#L166).

Just to be sure I specifically tested with both openOutputStream and uploadBlob with a large blob which hits the outputStream as well, and the code you have above with a small upload which does not use the stream. In all of these timeout was set.

I do have a couple guesses as to the confusion below:

This is a per request timeout. Do note that for large blobs (great than the singleBlobPutThresholdInBytes, also on the BlobRequestOptions) more than one service request can be made. In this case you'd want to look at the maximumExecutionTimeSetting.

In the example you give you're setting timeout interval so low it's being rounded to 0. Since this doesn't make sense (no request could ever complete in 1ms), the service it ignores it. If you set the timeout interval to something more reasonable (ie 5000 ms = 5s) you'll see more expected behavior.

**Comment by dadoonet** [comment]
I probably missed something. I'll try again and will update (or close) this ticket.

**Comment by dadoonet** [comment]
So I created this test:

```java
    public void testDumb() throws URISyntaxException, StorageException, IOException, InvalidKeyException {
        String connectionString = "MY-AZURE-STRING";

        CloudStorageAccount storageAccount = CloudStorageAccount.parse(connectionString);
        CloudBlobClient client = storageAccount.createCloudBlobClient();
        client.getDefaultRequestOptions().setTimeoutIntervalInMs(1000);
        CloudBlobContainer container = client.getContainerReference("dumb");
        container.createIfNotExists();
        CloudBlockBlob blob = container.getBlockBlobReference("blob");


        File sourceFile = File.createTempFile("sourceFile", ".tmp");

        try {
            int fileSize = 10000000;

            byte[] buffer = new byte[fileSize];
            Random random = new Random();
            random.nextBytes(buffer);

            logger.info("Generate local file");
            FileOutputStream fos = new FileOutputStream(sourceFile);
            fos.write(buffer);
            fos.close();
            logger.info("End generate local file");

            FileInputStream fis = new FileInputStream(sourceFile);

            logger.info("Start uploading");
            blob.upload(fis, fileSize);
            logger.info("End uploading");

        }
        finally {
            if (sourceFile.exists()) {
                sourceFile.delete();
            }
        }
```

As you can see, I'm setting a 1s timeout here.

Then I upload a big file (10 mb). Not that big but with my home bandwidth that's big enough :)

```
[2016-01-16 11:31:58,546][INFO ][org.elasticsearch.cloud.azure.storage] Generate local file
[2016-01-16 11:31:58,584][INFO ][org.elasticsearch.cloud.azure.storage] End generate local file
[2016-01-16 11:31:58,584][INFO ][org.elasticsearch.cloud.azure.storage] Start uploading
[2016-01-16 11:33:21,543][INFO ][org.elasticsearch.cloud.azure.storage] End uploading
```

Although the upload took 1 minute and 23 seconds, I never get any timeout here.

But may be that's not the right way to reproduce the "issue"? Or my assumption about what is the timeout is wrong here?

Thanks for your help so far.

**Comment by craigwi** [comment]
@dadoonet, I also created a test program and confirmed the lack of timeout parameter on the url.  However, I discovered the issue: we have been using azure storage library version 2.0.0; this bug was fixed in version 3.0.0 released last August.

**Comment by dadoonet** [comment]
Thanks @craigwi 

I just updated my test to 4.0.0 (which I planned at https://github.com/elastic/elasticsearch/issues/15976) but I'm not getting any timeout.
Were you able on your end to check with the same code that it was solved?

Thanks!

**Comment by craigwi** [comment]
I checked three things.  First, in v2.0.0 of BaseRequest.java the line "builder.add(TIMEOUT, String.valueOf(options.getTimeoutIntervalInMs() / 1000));" was AFTER "final URL resourceUrl = builder.addToURI(uri).toURL();".  In v3.0.0 and later the relative order of those two lines are reversed.

Second, I created a simple test program, based off the code you wrote above and using v3.0.0, which actually uploaded a string to Azure.  I stepped through the code and saw the URL being created properly. 

Third, I ran netmon and captured the output of my test program; the "&timeout=10" added to the URL on each of the three HTTP requests (HEAD, PUT content, PUT block list).  

I did not, however, verify that Azure actually timed out.

Craig.

**Comment by dadoonet** [comment]
Thanks @craigwi for this. I read again the Javadoc and I understand now what timeout actually means. It's a server side timeout and not client side timeout. It starts when the client has finished uploading the data and ends when the server has finished storing it on Azure storage.

I tested 3.1.0 as suggested in https://github.com/elastic/elasticsearch/issues/12567#issuecomment-172733697.

```
java.io.IOException
	at __randomizedtesting.SeedInfo.seed([91BC11AEF16E073F:6886FA5308FCE4D8]:0)
	at com.microsoft.azure.storage.core.Utility.initIOException(Utility.java:643)
	at com.microsoft.azure.storage.blob.BlobOutputStream.writeBlock(BlobOutputStream.java:444)
	at com.microsoft.azure.storage.blob.BlobOutputStream.access$000(BlobOutputStream.java:53)
	at com.microsoft.azure.storage.blob.BlobOutputStream$1.call(BlobOutputStream.java:388)
	at com.microsoft.azure.storage.blob.BlobOutputStream$1.call(BlobOutputStream.java:385)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.microsoft.azure.storage.StorageException: Operation could not be completed within the specified time.
	at com.microsoft.azure.storage.StorageException.translateException(StorageException.java:89)
	at com.microsoft.azure.storage.core.StorageRequest.materializeException(StorageRequest.java:305)
	at com.microsoft.azure.storage.core.ExecutionEngine.executeWithRetry(ExecutionEngine.java:175)
	at com.microsoft.azure.storage.blob.CloudBlockBlob.uploadBlockInternal(CloudBlockBlob.java:1006)
	at com.microsoft.azure.storage.blob.CloudBlockBlob.uploadBlock(CloudBlockBlob.java:978)
	at com.microsoft.azure.storage.blob.BlobOutputStream.writeBlock(BlobOutputStream.java:438)
	... 9 more
```

I also test it with 4.0.0 and it looks fine to me. I don't understand why it did not fail [on last attempt](https://github.com/Azure/azure-storage-java/issues/71#issuecomment-172513724).

So I'm closing this issue as it has been fixed somewhere between 2.0.0 and 3.1.0.

I will open another issue to ask if we can have a client side timeout as well which would be super handy for our use case. Not sure if it's doable though.

On a side note, I found that when we hit a timeout on the server, the response (exception) comes a long time after the full blob has been fully uploaded. With no timeout set the response comes really faster.
